core/text/src/markdown/lexer.rs

```diff
@@ -0,0 +1,218 @@
+    RBrace,
+    Space(String),
+    Word(String),
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct Tokens<'a> {
+    pub tok: &'a [LexToken],
+    start: usize,
+    end: usize,
+}
+
+impl<'a> Tokens<'a> {
+    #[allow(clippy::ptr_arg)]
+    pub fn new(vec: &'a Vec<LexToken>) -> Tokens<'a> {
+        Tokens {
+            tok: &vec[..],
+            start: 0,
+            end: vec.len(),
+        }
+    }
+}
+
+impl<'a> InputLength for Tokens<'a> {
+    fn input_len(&self) -> usize {
+        self.tok.len()
+    }
+}
+
+impl<'a> InputTake for Tokens<'a> {
+    fn take(&self, count: usize) -> Self {
+        Tokens {
+            tok: &self.tok[..count],
+            start: 0,
+            end: count,
+        }
+    }
+
+    fn take_split(&self, count: usize) -> (Self, Self) {
+        let (prefix, suffix) = self.tok.split_at(count);
+        let first = Tokens {
+            tok: prefix,
+            start: 0,
+            end: prefix.len(),
+        };
+        let second = Tokens {
+            tok: suffix,
+            start: 0,
+            end: suffix.len(),
+        };
+
+        (second, first)
+    }
+}
+
+impl<'a> InputLength for LexToken {
+    fn input_len(&self) -> usize {
+        1
+    }
+}
+
+impl<'a> Slice<Range<usize>> for Tokens<'a> {
+    fn slice(&self, range: Range<usize>) -> Self {
+        Tokens {
+            tok: &self.tok[range.clone()],
+            start: self.start + range.start,
+            end: self.end + range.end,
+        }
+    }
+}
+
+impl<'a> Slice<RangeTo<usize>> for Tokens<'a> {
+    fn slice(&self, range: RangeTo<usize>) -> Self {
+        self.slice(0..range.end)
+    }
+}
+
+impl<'a> Slice<RangeFrom<usize>> for Tokens<'a> {
+    fn slice(&self, range: RangeFrom<usize>) -> Self {
+        self.slice(range.start..self.end - self.start)
+    }
+}
+
+impl<'a> Slice<RangeFull> for Tokens<'a> {
+    fn slice(&self, _: RangeFull) -> Self {
+        Tokens {
+            tok: self.tok,
+            start: self.start,
+            end: self.end,
+        }
+    }
+}
+
+impl<'a> InputIter for Tokens<'a> {
+    type Item = &'a LexToken;
+    type Iter = Enumerate<Iter<'a, LexToken>>;
+    type IterElem = Iter<'a, LexToken>;
+
+    fn iter_indices(&self) -> Self::Iter {
+        self.tok.iter().enumerate()
+    }
+
+    fn iter_elements(&self) -> Self::IterElem {
+        self.tok.iter()
+    }
+
+    fn position<P>(&self, predicate: P) -> Option<usize>
+    where
+        P: Fn(Self::Item) -> bool,
+    {
+        self.tok.iter().position(|b| predicate(b))
+    }
+
+    fn slice_index(&self, count: usize) -> Option<usize> {
+        if self.tok.len() >= count {
+            Some(count)
+        } else {
+            None
+        }
+    }
+}
+
+pub fn lex_control_word(input: &str) -> IResult<&str, LexToken> {
+    map(tag("@"), |_| LexToken::ControlWordStarter)(input)
+}
+
+pub fn lex_spaces(input: &str) -> IResult<&str, LexToken> {
+    map(space1, |s: &str| LexToken::Space(s.to_string()))(input)
```

Comment:
We could reduce allocations by storing string slices rather than owned `String`s, though it unlikely matters in practice. It seems fairly trivial to do in this case, however.